lib/ansible/plugins/callback/junit.py

```diff
@@ -0,0 +1,237 @@
+except ImportError:
+    HAS_JUNIT_XML = False
+
+
+class CallbackModule(CallbackBase):
+    """
+    This callback writes playbook output to a JUnit formatted XML file.
+
+    Tasks show up in the report as follows:
+        'ok': pass
+        'failed' with 'EXPECTED FAILURE' in the task name: pass
+        'failed' due to an exception: error
+        'failed' for other reasons: failure
+        'skipped': skipped
+
+    This plugin makes use of the following environment variables:
+        JUNIT_OUTPUT_DIR (optional): Directory to write XML files to.
+                                     Default: Current working directory.
+
+    Requires:
+        junit_xml
+
+    """
+
+    CALLBACK_VERSION = 2.0
+    CALLBACK_TYPE = 'aggregate'
+    CALLBACK_NAME = 'junit'
+    CALLBACK_NEEDS_WHITELIST = True
+
+    def __init__(self):
+        super(CallbackModule, self).__init__()
+
+        self._output_dir = os.getenv('JUNIT_OUTPUT_DIR', os.getcwd())
+        self._playbook_path = None
+        self._playbook_name = None
+        self._play_name = None
+        self._task_data = OrderedDict()
+
+        self.disabled = False
+
+        if not HAS_JUNIT_XML:
+            self.disabled = True
+            self._display.warning('The `junit_xml` python module is not installed. '
+                                  'Disabling the `junit` callback plugin.')
+
+    def _start_task(self, task):
+        """ record the start of a task for one or more hosts """
+
+        uuid = task._uuid
+
+        if uuid in self._task_data:
+            return
+
+        play = self._play_name
+        name = task.get_name().strip()
+        path = task.get_path()
+
+        if not task.no_log:
+            args = ', '.join(('%s=%s' % a for a in task.args.items()))
+            if args:
+                name += ' ' + args
+
+        self._task_data[uuid] = TaskData(uuid, name, path, play)
+
+    def _finish_task(self, status, result):
+        """ record the results of a task for a single host """
+
+        task_uuid = result._task._uuid
+
+        if hasattr(result, '_host'):
+            host_uuid = result._host._uuid
+            host_name = result._host.name
+        else:
+            host_uuid = 'include'
+            host_name = 'include'
+
+        task_data = self._task_data[task_uuid]
+
+        if status == 'failed' and 'EXPECTED FAILURE' in task_data.name:
+            status = 'ok'
+
+        task_data.add_host(HostData(host_uuid, host_name, status, result))
+
+    def _build_test_case(self, task_data, host_data):
+        """ build a TestCase from the given TaskData and HostData """
+
+        name = '[%s] %s: %s' % (host_data.name, task_data.play, task_data.name)
+        duration = host_data.finish - task_data.start
+
+        if host_data.status == 'included':
+            return TestCase(name, task_data.path, duration, host_data.result)
+
+        res = host_data.result._result
+        rc = res.get('rc', 0)
+        dump = self._dump_results(res, indent=0)
```

Comment:
Aside from self._dump_results here, this could almost be a static method or module scope method.

I sort of want to suggest always just passing the entire host_data.result to TestCase. Would seem more useful for the intended use case  (viewing results of our test cases).  But, I could also see this being useful with real 'live' data 
and the results being sensitive  (something like someone using ansible via jenkins as part of a test setup, and getting the task results as jenkins would be handy). Maybe include an AbridgedTestCase(TestCase) that can do the _dump_results style cleanup itself?

Something that has a TestCase() [or perhaps as a classmethod alternative constructor] could also handle figuring out the duration, name, and add_*_info info.

Then this method is more or less
```
def _build_test_case(self, task_data, host_data):
    return AnsibleTestCase(task_data, host_data)
```

Similarly for TestSuite.
test_suite = AnsibleTestSuite(playbook_name, task_data)

(And maybe TaskData would get playbook_name passed to it's init?)